### PR TITLE
[DON'T SQUASH] LLVM update 9/21/21

### PR DESCRIPTION
**Pull request by sjw36**

### CODE REVIEW

**Comment by whchung** [comment]
@sjw36 public facing CI (ml-ci.amd.com) is failing due to node issue it seems. do you want to restart the test or is there another job in the queue already?

**Comment by sjw36** [comment]
I'm on it..

**Comment by whchung** [comment]
@jerryyin / @yiqian1 / @okakarpa According to the logs, the latest CI run for this PR is stuck at waiting to find an available gfx908 node. Wondering how to move it forward? I do want to ensure all resnet50 configs can be validated before we take this PR.

**Comment by jerryyin** [comment]
@whchung It looks to me that @sjw36 just kicked out a new static lib test and I don't see public CI failing.

On the other hand, a previous run log is good enough for me: http://rocmhead.amd.com:8080/blue/organizations/jenkins/mlir/detail/PR-396/9/pipeline. I'm giving approval based on this "almost" all pass of MIOpen integration tests.

**Review by jerryyin — APPROVED** [review]

**Comment by sjw36** [comment]
Since I am still trying to get MIOpenDriver working locally (slowed down today on ixt-rack-112)... so I went ahead and kicked off another run of nightly testing.

**Comment by whchung** [comment]
> On the other hand, a previous run log is good enough for me: http://rocmhead.amd.com:8080/blue/organizations/jenkins/mlir/detail/PR-396/9/pipeline. I'm giving approval based on this "almost" all pass of MIOpen integration tests.

Thanks for letting me know. Run 9 does seem to be "almost fine". I think I have a larger confidence taking it.

**Comment by sjw36** [comment]
The one "failure" also had passed many configs before the node died.  Shall we go ahead and merge this and see how nightly runs tonight?

**Comment by sjw36** [comment]
> Thanks for letting me know. Run 9 does seem to be "almost fine". I think I have a larger confidence taking it.

@whchung do you approve also?

**Comment by whchung** [comment]
> @whchung do you approve also?

We don't have any understanding nor a solution to `pthread_create failed: Resource temporarily unavailable` yet so I can't give my approval.

I can see it blocking us from getting a stable nightly CI. I would like to see a better investigation about the issue before we can safely merge the PR.

10.236.106.97:8080/blue/organizations/jenkins/mlir/detail/PR-396/10/pipeline/1243

is an example of this PR fails again due to `pthread_create`. Notice when the pipeline runs it's supposed to be have exclusive access to a node.


Plus, we have #385 currently blocking nightly CI as well. I filed #401 yesterday but it doesn't seem to address all the issues yet. None of these 2 PRs are in this PR yet.

**Comment by whchung** [comment]
@sjw36 one idea I'd like to propose to experiment is to try limit the workers in LLVM lit test, as well as limiting threads spawned within MLIR implementation. I don't exactly know how to do them though so I think some studies / experiments may be needed.

**Comment by jerryyin** [comment]
@whchung Good catch on that `LLVM ERROR: pthread_create failed`! Yes, I agree that unless we can prove this issue is solved, this shouldn't be merged otherwise it just becomes a pain in nightly regression and we still have to solve it.

**Comment by sjw36** [comment]
It doesn't seem to be a problem with rocm-runner so I just disabled multi-threading in libMLIRMIOpen.

File: mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp [hunk 1]
```diff
@@ -34,6 +34,7 @@ struct MiirHandle_s {
     return registry;
   }
   MiirHandle_s() : context(initRegistry()) {
+    context.disableMultithreading();
```
Comment (jerryyin) [inline mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp]: @sjw36 Please file a ticket to specifically triage and remove this line later. Unlike the public facing mutex that disallow user multi-threading, this has the potential to degrade the lowering speed.

Comment (sjw36) [reply mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp]: See ticket https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/353

Comment (jerryyin) [reply mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp]: Thank you!

**Comment by krzysz00** [comment]
One observation I have re threads is that, when I've debugged mlir-rocm-runner, it spawns what are, by all accounts, 64 threads. So, having a lot of mlir-rocm-runners active at once could be causing the process table to run out of entries.

**Comment by whchung** [comment]
jenkins: retest this please.

**Comment by okakarpa** [comment]
> jenkins: retest this please.

@whchung  This PR seems to be hanged at build-5 could be the reason your above comments are not working i am not sure, irrespective of it seems like the build is hung for a day or so is this expected ?

http://ml-ci.amd.com:21096/job/MLIR/job/mlir/job/PR-396/5/console

**Comment by whchung** [comment]
@sjw36 per observation from @okakarpa , suggest we abort build 5? I don't know if that build have your logic to disable mult-threading or not.

http://ml-ci.amd.com:21096/job/MLIR/job/mlir/job/PR-396/5/console

**Comment by sjw36** [comment]
@whchung it looks like it is already aborted. If not please abort.

**Comment by whchung** [comment]
@sjw36 I noticed you haven't cherry-picked the commit I recommended on #401:

```
@sjw36 You might want to cherry pick commits in this PR (specifically d6bd4c1) to #396 . We want GNU parallel to run with -j 1 to suppress CPU threads from spawning unbounded. I believe this should address some issues we see running tests from MIOpen side.

```
The commit is now already in `miopen-dialect` yesterday after invoked my admin privilege to commit arbitrary commits to the mainline. For this PR though I think you should do that too to be on the safe side.

**Comment by whchung** [comment]
In build 33(!), "Shared library build an test" has finally passed. Keeping an eye on "Random E2E test" as well as "MIOpen resnet50 config test"...

**Comment by whchung** [comment]
@sjw36 build 33 ran to completion but failed. I would recommend you take a look at CI build log to figure out what happened. We can see that "Shared library build and test" and "Random E2E test" stages passed so at least all unit tests supplied in the repository are passing. I would assume the failures are coming from driving from MIOpen.

File: mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp [hunk 2]
```diff
@@ -29,9 +29,13 @@
 
 namespace {
 struct MiirHandle_s {
-  MiirHandle_s() {
+  DialectRegistry initRegistry() {
+    DialectRegistry registry;
+    registerAllDialects(registry);
+    return registry;
```
Comment (jerryyin) [inline mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp]: Just to capture our discussion:

I don't feel like `registerAllDialects` is thread safe therefore we should:
 - Move it to `miirLazyInit`
 - Make `registry` a static variable and initialized only in `miirLazyInit`
 - At each handle creation, context can initialize from that single registry

**Pull request merged**